File: project/src/main/java/euphony/lib/transmitter/EuDataEncoder.java
package euphony.lib.transmitter;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.Arrays;

import euphony.lib.util.EuCodec;

public class EuDataEncoder extends EuCodec {
	private String mOriginalSource;

	EuDataEncoder() {
	}

	public EuDataEncoder(String _source) {
		mOriginalSource = _source;
	}

	public String encodeHexCharSource() {
		return encodeStaticHexCharSource(mOriginalSource);
	}

	public static String encodeStaticBinaryCharSource(String _source){
		StringBuilder strBuilder = new StringBuilder();

		_source = encodeStaticHexCharSource(_source);

		for (int i = 0; i < _source.length(); i++) {
			int data = Integer.parseInt(_source.charAt(i) + "", 16);
			strBuilder.append(String.format("%04d", Integer.parseInt(Integer.toBinaryString(data))));
		}

		return strBuilder.toString();
	}

	public static String encodeStaticHexCharSource(String _source) {
		StringBuilder strBuilder = new StringBuilder();

		for (int i = 0; i < _source.length(); i++) {
			int data =_source.charAt(i);
			strBuilder.append(Integer.toHexString(data));
		}

		return strBuilder.toString();
	}

	/*
	TODO: 40 BASE should be implemented.
	public static byte[] base40StaticEncode(String _source)
	{
		try {
			ByteArrayOutputStream baos = new ByteArrayOutputStream();
			DataOutputStream dos = new DataOutputStream(baos);
			for(int i = 0; i < _source.length(); i+=3)
			{
				switch(Math.min(3, _source.length() - i)) {
				case 1:
					byte b = base40Index[_source.charAt(i)];
					dos.writeByte(b);
					break;
				case 2:
					char ch = (char) ((base40Index[_source.charAt(i+1)]) * 40 + base40Index[_source.charAt(i)]);
					dos.writeChar(ch);
					break;
				case 3:
					char ch2 = (char) ((char) ((base40Index[_source.charAt(i+2)]) * 40 + base40Index[_source.charAt(i+1)]) * 40 + base40Index[_source.charAt(i)]);
					dos.writeChar(ch2);
					
					break;
				}
			}
			return baos.toByteArray();
		} 
		catch(IOException e) 
		{
			throw new AssertionError(e);
		}
	}
	*/

	public String getOriginalSource() {
		return mOriginalSource;
	}
	public void setOriginalSource(String mOriginalSource) {
		this.mOriginalSource = mOriginalSource;
	}
}

File: project/src/main/kotlin/euphony/lib/transmitter/EuDataEncoder.kt
package euphony.lib.transmitter

import android.util.Log
import euphony.lib.util.EuCodec
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.io.IOException

class EuDataEncoder : EuCodec {
    var originalSource: String? = null

    internal constructor() {}
    constructor(_source: String) {
        originalSource = _source
    }

    fun encodeHexCharSource(): String {
        return encodeStaticHexCharSource(originalSource)
    }

    fun encodeStaticHexCharSource(_source: String?): String {
        val strBuilder = StringBuilder()

        for (i in 0 until _source!!.length) {
            val data = _source[i].toInt()
            strBuilder.append(Integer.toHexString(data))
        }

        return strBuilder.toString()
    }

    companion object {

        fun base40StaticEncode(_source: String): ByteArray {
            try {
                val baos = ByteArrayOutputStream()
                val dos = DataOutputStream(baos)
                var i = 0
                while (i < _source.length) {
                    when (Math.min(3, _source.length - i)) {
                        1 -> {
                            val b = EuCodec.base40Index[_source[i] as Int]
                            dos.writeByte(b.toInt())
                        }
                        2 -> {
                            val ch = (EuCodec.base40Index[_source[i + 1] as Int] * 40 + EuCodec.base40Index[_source[i] as Int]).toChar()
                            dos.writeChar(ch.toInt())
                        }
                        3 -> {
                            val ch2 = ((EuCodec.base40Index[_source[i + 2] as Int] * 40 + EuCodec.base40Index[_source[i + 1] as Int]).toChar().toInt() * 40 + EuCodec.base40Index[_source[i] as Int]).toChar()
                            dos.writeChar(ch2.toInt())
                        }
                    }
                    i += 3
                }
                return baos.toByteArray()
            } catch (e: IOException) {
                throw AssertionError(e)
            }

        }
    }
}